File: shell/src/dev/symbiosis/kenji/CoverArt.java
package dev.symbiosis.kenji;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Covers without touching libkenjinx in the launcher process.
 *
 * Order: cache → sidecar jpg → Eden/Kenji icon folders → NSP Control NCA.
 * A missing cover is a placeholder, never a crash.
 */
public final class CoverArt {
    private CoverArt() {}

    public static File cacheFile(Context context, GameItem game) {
        File dir = new File(context.getCacheDir(), "covers");
        dir.mkdirs();
        String id = game.titleId;
        if (id == null || id.isEmpty()) {
            id = Integer.toHexString(game.key().hashCode());
        }
        return new File(dir, id + ".jpg");
    }

    public static Bitmap load(Context context, GameItem game) {
        if (game == null) return null;
        File cache = cacheFile(context, game);
        if (cache.isFile() && cache.length() > 32) {
            Bitmap b = BitmapFactory.decodeFile(cache.getAbsolutePath());
            if (b != null) return b;
        }
        File found = findExisting(context, game);
        if (found != null) {
            copy(found, cache);
            return BitmapFactory.decodeFile(found.getAbsolutePath());
        }
        byte[] jpeg = extractNsp(context, game.path);
        if (jpeg != null && jpeg.length > 64) {
            write(cache, jpeg);
            return BitmapFactory.decodeByteArray(jpeg, 0, jpeg.length);
        }
        return null;
    }

    private static File findExisting(Context context, GameItem game) {
        String tid = game.titleId == null ? "" : game.titleId.toLowerCase(Locale.US);
        String tidU = game.titleId == null ? "" : game.titleId.toUpperCase(Locale.US);
        File rom = game.path.startsWith("/") ? new File(game.path) : null;
        if (rom != null && rom.isFile()) {
            String stem = rom.getName();
            int dot = stem.lastIndexOf('.');
            if (dot > 0) stem = stem.substring(0, dot);
            File parent = rom.getParentFile();
            if (parent != null) {
                String[] names = {
                        stem + ".jpg", stem + ".png", stem + ".jpeg",
                        tidU + ".jpg", tidU + ".png", tid + ".jpg", tid + ".png",
                        "cover.jpg", "cover.png", "icon.jpg", "icon.png",
                        "folder.jpg", "folder.png"
                };
                File walk = parent;
                for (int up = 0; up < 3 && walk != null; up++) {
                    for (String n : names) {
                        File f = new File(walk, n);
                        if (f.isFile() && f.length() > 32) return f;
                    }
                    File[] kids = walk.listFiles();
                    if (kids != null) {
                        String lowStem = stem.toLowerCase(Locale.US);
                        for (File k : kids) {
                            String kn = k.getName().toLowerCase(Locale.US);
                            if (!k.isFile() || k.length() <= 32) continue;
                            boolean img = kn.endsWith(".jpg") || kn.endsWith(".jpeg") || kn.endsWith(".png");
                            if (!img) continue;
                            if ((!tid.isEmpty() && kn.contains(tid))
                                    || kn.contains(lowStem)
                                    || kn.contains("chimera")
                                    || kn.contains("cover")
                                    || kn.contains("icon")) {
                                return k;
                            }
                        }
                    }
                    walk = walk.getParentFile();
                }
            }
        }
        File[] roots = new File[]{
                DataSeed.appPath(context),
                DataSeed.bestEden(context),
                DataSeed.userRoot(context),
                new File(android.os.Environment.getExternalStorageDirectory(),
                        "Android/data/org.kenjinx.android/files"),
                new File(android.os.Environment.getExternalStorageDirectory(),
                        "Android/data/dev.eden.eden_emulator/files"),
                new File(android.os.Environment.getExternalStorageDirectory(),
                        "Android/data/org.yuzu.yuzu_emu/files")
        };
        String[] rel = {
                "cache/game_list/" + tid + ".png",
                "cache/game_list/" + tidU + ".png",
                "cache/game_list/" + tid + ".jpg",
                "cache/game_list/" + tidU + ".jpg",
                "cache/" + tid + ".png",
                "cache/" + tidU + ".jpg",
                "cache/icons/" + tid + ".png",
                "cache/icons/" + tidU + ".jpg",
                "games/" + tidU + "/icon.jpg",
                "games/" + tidU + "/icon.png",
                "games/" + tid + "/icon.jpg",
                "nand/cache/" + tid + ".png",
                "icons/" + tid + ".png",
                "icons/" + tidU + ".jpg"
        };
        for (File root : roots) {
            if (root == null) continue;
            for (String r : rel) {
                File f = new File(root, r);
                if (f.isFile() && f.length() > 32) return f;
            }
        }
        return null;
    }

    private static byte[] extractNsp(Context context, String path) {
        if (path == null || !path.startsWith("/")) return null;
        File rom = new File(path);
        if (!rom.isFile()) return null;
        String low = rom.getName().toLowerCase(Locale.US);
        if (!low.endsWith(".nsp")) return null;
        File keys = new File(DataSeed.appPath(context), "system/prod.keys");
        Map<String, byte[]> map = readKeys(keys);
        byte[] headerKey = map.get("header_key");
        if (headerKey == null || headerKey.length != 32) return null;
        RandomAccessFile raf = null;
        try {
            raf = new RandomAccessFile(rom, "r");
            byte[] magic = new byte[4];
            raf.readFully(magic);
            if (!"PFS0".equals(new String(magic, StandardCharsets.US_ASCII))) return null;
            int count = Integer.reverseBytes(raf.readInt());
            int strSize = Integer.reverseBytes(raf.readInt());
            raf.readInt();
            if (count <= 0 || count > 4096 || strSize <= 0 || strSize > 1_000_000) return null;
            long[] offs = new long[count];
            long[] sizes = new long[count];
            int[] nameOff = new int[count];
            for (int i = 0; i < count; i++) {
                offs[i] = Long.reverseBytes(raf.readLong());
                sizes[i] = Long.reverseBytes(raf.readLong());
                nameOff[i] = Integer.reverseBytes(raf.readInt());
                raf.readInt();
            }
            byte[] strings = new byte[strSize];
            raf.readFully(strings);
            long dataStart = 16L + 24L * count + strSize;
            for (int i = 0; i < count; i++) {
                String name = cstr(strings, nameOff[i]).toLowerCase(Locale.US);
                if (!name.endsWith(".nca") || name.endsWith(".cnmt.nca")) continue;
                if (sizes[i] < 0xC00 || sizes[i] > 12L * 1024 * 1024) continue;
                byte[] nca = new byte[(int) sizes[i]];
                raf.seek(dataStart + offs[i]);
                raf.readFully(nca);
                byte[] jpeg = controlIcon(nca, map, headerKey);
                if (jpeg != null) return jpeg;
            }
        } catch (Throwable ignored) {
        } finally {
            if (raf != null) try { raf.close(); } catch (Exception ignored) {}
        }
        return null;
    }

    private static byte[] controlIcon(byte[] nca, Map<String, byte[]> keys, byte[] headerKey) {
        try {
            byte[] header = decryptXts(headerKey, slice(nca, 0, 0xC00));
            if (header.length < 0x340) return null;
            if (header[0x200] != 'N' || header[0x201] != 'C' || header[0x202] != 'A') return null;
            int contentType = header[0x205] & 0xFF;
            if (contentType != 2) return null; // Control
            boolean hasRights = false;
            for (int i = 0; i < 16; i++) {
                if (header[0x230 + i] != 0) { hasRights = true; break; }
            }
            if (hasRights) return null; // needs title.keys
            int keyGen = header[0x220] & 0xFF;
            if (keyGen == 0) keyGen = header[0x206] & 0xFF;
            int kaIndex = header[0x207] & 0xFF;
            String kaName = kaIndex == 1 ? "ocean" : kaIndex == 2 ? "system" : "application";
            String kaKey = String.format(Locale.US, "key_area_key_%s_%02x", kaName, keyGen);
            byte[] kak = keys.get(kaKey);
            if (kak == null || kak.length != 16) return null;
            byte[] decKeys = aesEcb(kak, slice(header, 0x300, 0x40), false);
            if (decKeys.length < 48) return null;
            byte[] sectionKey = slice(decKeys, 32, 16);
            int startSector = le32(header, 0x240);
            int endSector = le32(header, 0x244);
            if (endSector <= startSector) return null;
            int start = startSector * 0x200;
            int size = (endSector - startSector) * 0x200;
            if (start < 0 || size <= 0 || start + size > nca.length) return null;
            if (size > 8 * 1024 * 1024) size = 8 * 1024 * 1024;
            byte[] ctr = new byte[16];
            // FS header lives at the start of the section; CTR field at +0x140 after decrypt.
            // First decrypt a window with zeroed offset, then refine.
            byte[] section = aesCtr(sectionKey, ctr, slice(nca, start, size), 0);
            // After decrypt, bytes at 0x140 are the stored CTR. Re-decrypt if needed.
            if (section.length > 0x148) {
                byte[] stored = slice(section, 0x140, 8);
                boolean nonzero = false;
                for (byte b : stored) if (b != 0) nonzero = true;
                if (nonzero) {
                    System.arraycopy(stored, 0, ctr, 0, 8);
                    section = aesCtr(sectionKey, ctr, slice(nca, start, size), 0);
                }
            }
            return findJpeg(section);
        } catch (Throwable t) {
            return null;
        }
    }

    private static byte[] findJpeg(byte[] data) {
        if (data == null) return null;
        int start = -1;
        for (int i = 0; i < data.length - 3; i++) {
            if ((data[i] & 0xFF) == 0xFF && (data[i + 1] & 0xFF) == 0xD8 && (data[i + 2] & 0xFF) == 0xFF) {
                start = i;
                break;
            }
        }
        if (start < 0) return null;
        for (int i = start + 3; i < data.length - 1; i++) {
            if ((data[i] & 0xFF) == 0xFF && (data[i + 1] & 0xFF) == 0xD9) {
                int len = i + 2 - start;
                if (len < 64 || len > 600_000) continue;
                byte[] out = new byte[len];
                System.arraycopy(data, start, out, 0, len);
                return out;
            }
        }
        return null;
    }

    private static Map<String, byte[]> readKeys(File prod) {
        Map<String, byte[]> map = new HashMap<String, byte[]>();
        if (prod == null || !prod.isFile()) return map;
        try {
            java.util.Scanner s = new java.util.Scanner(prod, "UTF-8");
            while (s.hasNextLine()) {
                String line = s.nextLine().trim();
                if (line.isEmpty() || line.startsWith("#") || !line.contains("=")) continue;
                int eq = line.indexOf('=');
                String k = line.substring(0, eq).trim().toLowerCase(Locale.US);
                String hex = line.substring(eq + 1).trim().replace(" ", "");
                if (hex.length() < 32 || hex.length() % 2 != 0) continue;
                boolean ok = true;
                for (int i = 0; i < hex.length(); i++) {
                    char c = hex.charAt(i);
                    if (Character.digit(c, 16) < 0) { ok = false; break; }
                }
                if (!ok) continue;
                byte[] b = new byte[hex.length() / 2];
                for (int i = 0; i < b.length; i++) {
                    b[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
                }
                map.put(k, b);
            }
            s.close();
        } catch (Exception ignored) {
        }
        return map;
    }

    private static byte[] decryptXts(byte[] key, byte[] data) throws Exception {
        SecretKeySpec k1 = new SecretKeySpec(key, 0, 16, "AES");
        SecretKeySpec k2 = new SecretKeySpec(key, 16, 16, "AES");
        Cipher dec = Cipher.getInstance("AES/ECB/NoPadding");
        Cipher twk = Cipher.getInstance("AES/ECB/NoPadding");
        dec.init(Cipher.DECRYPT_MODE, k1);
        twk.init(Cipher.ENCRYPT_MODE, k2);
        byte[] out = new byte[data.length];
        int sector = 0;
        int off = 0;
        final int SECTOR = 0x200;
        while (off < data.length) {
            byte[] tweak = new byte[16];
            long v = sector;
            for (int i = 0; i < 16; i++) {
                tweak[i] = (byte) (v & 0xFF);
                v >>>= 8;
            }
            byte[] t = twk.doFinal(tweak);
            int i = 0;
            while (i < SECTOR && off + i < data.length) {
                byte[] block = new byte[16];
                System.arraycopy(data, off + i, block, 0, 16);
                xor(block, t);
                byte[] plain = dec.doFinal(block);
                xor(plain, t);
                System.arraycopy(plain, 0, out, off + i, 16);
                gf128(t);
                i += 16;
            }
            off += SECTOR;
            sector++;
        }
        return out;
    }

    private static byte[] aesEcb(byte[] key, byte[] data, boolean encrypt) throws Exception {
        Cipher c = Cipher.getInstance("AES/ECB/NoPadding");
        c.init(encrypt ? Cipher.ENCRYPT_MODE : Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"));
        return c.doFinal(data);
    }

    private static byte[] aesCtr(byte[] key, byte[] ctr0, byte[] data, long offset) throws Exception {
        byte[] ctr = ctr0.clone();
        long blocks = offset / 16;
        for (int i = 15; i >= 8 && blocks > 0; i--) {
            long v = (ctr[i] & 0xFF) + (blocks & 0xFF);
            ctr[i] = (byte) v;
            blocks = (blocks >>> 8) + (v >>> 8);
        }
        Cipher c = Cipher.getInstance("AES/CTR/NoPadding");
        c.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"), new IvParameterSpec(ctr));
        return c.doFinal(data);
    }

    private static void xor(byte[] a, byte[] b) {
        for (int i = 0; i < a.length; i++) a[i] = (byte) (a[i] ^ b[i]);
    }

    private static void gf128(byte[] t) {
        int carry = 0;
        for (int i = 0; i < 16; i++) {
            int b = t[i] & 0xFF;
            t[i] = (byte) ((b << 1) | carry);
            carry = b >>> 7;
        }
        if (carry != 0) t[0] = (byte) (t[0] ^ 0x87);
    }

    private static int le32(byte[] b, int off) {
        return ByteBuffer.wrap(b, off, 4).order(ByteOrder.LITTLE_ENDIAN).getInt();
    }

    private static byte[] slice(byte[] src, int off, int len) {
        byte[] o = new byte[len];
        System.arraycopy(src, off, o, 0, len);
        return o;
    }

    private static String cstr(byte[] s, int off) {
        if (off < 0 || off >= s.length) return "";
        int e = off;
        while (e < s.length && s[e] != 0) e++;
        return new String(s, off, e - off, StandardCharsets.UTF_8);
    }

    private static void copy(File from, File to) {
        try {
            FileInputStream in = new FileInputStream(from);
            FileOutputStream out = new FileOutputStream(to);
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) out.write(buf, 0, n);
            in.close();
            out.close();
        } catch (Exception ignored) {
        }
    }

    private static void write(File to, byte[] data) {
        try {
            FileOutputStream out = new FileOutputStream(to);
            out.write(data);
            out.close();
        } catch (Exception ignored) {
        }
    }
}

File: shell/src/dev/symbiosis/kenji/CoverArt.kt
package dev.symbiosis.kenji

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.os.Environment
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors

/**
 * Icon without deviceGetGameInfo and without reading the whole NSP.
 * Order: disk cache → sidecar → Eden/Kenji icon files → Kenji view scrape
 * → native *Icon* (not GameInfo) → small PFS0 JPEG scan.
 */
object CoverArt {
    private val mem = ConcurrentHashMap<String, Bitmap>()
    private val failed = ConcurrentHashMap.newKeySet<String>()
    private val pool = Executors.newFixedThreadPool(2)

    fun cached(rom: RomList.Rom): Bitmap? = mem[key(rom)]

    fun load(context: Context, rom: RomList.Rom, done: (Bitmap?) -> Unit) {
        val k = key(rom)
        mem[k]?.let { done(it); return }
        if (k in failed) {
            done(null)
            return
        }
        val disk = diskFile(context, rom)
        if (disk.isFile && disk.length() > 200) {
            val bmp = decodeFile(disk)
            if (bmp != null) {
                mem[k] = bmp
                done(bmp)
                return
            }
        }
        pool.execute {
            val bmp = try {
                extract(context, rom)
            } catch (t: Throwable) {
                Log.w("KenjiSpace", "cover ${rom.file.name}", t)
                null
            }
            if (bmp != null) {
                mem[k] = bmp
                save(disk, bmp)
            } else {
                failed.add(k)
            }
            val act = context as? Activity
            if (act != null) act.runOnUiThread { done(bmp) } else done(bmp)
        }
    }

    private fun key(rom: RomList.Rom): String =
        rom.titleId.ifBlank { rom.file.absolutePath }

    private fun diskFile(context: Context, rom: RomList.Rom): File {
        val dir = File(DataSeed.playHome(context), "space_icons")
        val name = (rom.titleId.ifBlank { rom.file.nameWithoutExtension }).lowercase(Locale.US)
        return File(dir, "$name.jpg")
    }

    private fun extract(context: Context, rom: RomList.Rom): Bitmap? {
        sidecar(rom)?.let { return it }
        huntNamed(context, rom)?.let { return it }
        (context as? Activity)?.let { scrape(it, rom)?.let { b -> return b } }
        nativeIcon(rom)?.let { return it }
        val f = rom.file
        if (!f.isFile) return null
        val low = f.name.lowercase(Locale.US)
        if (low.endsWith(".nro")) return nroIcon(f)
        return pfs0Icon(f)
    }

    private fun sidecar(rom: RomList.Rom): Bitmap? {
        val parent = rom.file.parentFile ?: return null
        val base = rom.file.nameWithoutExtension
        val names = ArrayList<String>()
        names.add("$base.jpg"); names.add("$base.jpeg"); names.add("$base.png")
        if (rom.titleId.length >= 8) {
            names.add("${rom.titleId}.jpg")
            names.add("${rom.titleId}.png")
            names.add("${rom.titleId.lowercase(Locale.US)}.jpg")
            names.add("${rom.titleId.lowercase(Locale.US)}.png")
        }
        names.add("cover.jpg"); names.add("icon.jpg")
        for (n in names) {
            val f = File(parent, n)
            if (f.isFile && f.length() in 800L..900_000L) decodeFile(f)?.let { return it }
        }
        return null
    }

    private fun huntNamed(context: Context, rom: RomList.Rom): Bitmap? {
        val id = rom.titleId
        if (id.length < 8) return null
        val low = id.lowercase(Locale.US)
        val roots = ArrayList<File>()
        roots.add(DataSeed.playHome(context))
        DataSeed.edenDir(context)?.let { roots.add(File(it)) }
        val sd = Environment.getExternalStorageDirectory()
        listOf(
            "Android/data/dev.eden.eden_emulator/files",
            "Android/data/org.yuzu.yuzu_emu/files",
            "Android/data/org.sudachi.sudachi_emu/files",
            "Android/data/org.kenjinx.android/files",
            "Android/data/dev.symbiosis.kenji/files",
            "Download/ed/Eden/files",
            "Download/ed",
        ).forEach { rel ->
            val f = File(sd, rel)
            if (f.isDirectory) roots.add(f)
        }
        val rels = arrayOf(
            "space_icons/$low.jpg",
            "cache/game_list/$low.png",
            "cache/game_list/$low.jpg",
            "cache/icons/$low.png",
            "cache/icons/$low.jpg",
            "cache/$low.png",
            "cache/$low.jpg",
            "games/$low.jpg",
            "games/$low.png",
            "icons/$low.png",
            "icons/$low.jpg",
            "load/icons/$low.jpg",
            "sdmc/atmosphere/contents/$low/icon.jpg",
        )
        for (root in roots) {
            for (rel in rels) {
                val f = File(root, rel)
                if (f.isFile && f.length() in 800L..900_000L) decodeFile(f)?.let { return it }
            }
        }
        for (root in roots) {
            walkName(root, low, 0)?.let { return it }
        }
        return null
    }

    private fun walkName(dir: File, needle: String, depth: Int): Bitmap? {
        if (depth > 4 || !dir.isDirectory) return null
        val kids = dir.listFiles() ?: return null
        var n = 0
        for (f in kids) {
            if (n++ > 80) break
            val name = f.name.lowercase(Locale.US)
            if (f.isFile && name.contains(needle) &&
                (name.endsWith(".jpg") || name.endsWith(".jpeg") || name.endsWith(".png")) &&
                f.length() in 800L..900_000L
            ) {
                decodeFile(f)?.let { return it }
            }
            if (f.isDirectory && !name.startsWith(".") && name != "registered" && name != "nand") {
                walkName(f, needle, depth + 1)?.let { return it }
            }
        }
        return null
    }

    private fun scrape(act: Activity, rom: RomList.Rom): Bitmap? {
        val want = rom.title.lowercase(Locale.US)
        var fallback: Bitmap? = null
        for (root in SpaceHook.allWindowsPublic()) {
            scrapeWalk(root, 0, want) { bmp ->
                if (bmp.width >= 96 && bmp.height >= 96 && bmp.width <= 1024) {
                    fallback = bmp
                }
            }
        }
        return fallback
    }

    private fun scrapeWalk(v: View, depth: Int, title: String, hit: (Bitmap) -> Unit) {
        if (depth > 16 || SpaceHook.isSpaceView(v)) return
        when (v) {
            is ImageView -> {
                val d = v.drawable
                val bmp = (d as? BitmapDrawable)?.bitmap
                if (bmp != null && !bmp.isRecycled) hit(bmp)
            }
            else -> {
                val d = v.background
                val bmp = (d as? BitmapDrawable)?.bitmap
                if (bmp != null && !bmp.isRecycled && bmp.width >= 96) hit(bmp)
            }
        }
        if (v is ViewGroup) {
            for (i in 0 until v.childCount) scrapeWalk(v.getChildAt(i), depth + 1, title, hit)
        }
    }

    /** Icon-only JNI. Never deviceGetGameInfo — that OOM'd the launcher. */
    private fun nativeIcon(rom: RomList.Rom): Bitmap? {
        val classes = arrayOf(
            "org.kenjinx.android.KenjinxNative",
            "org.kenjinx.android.native.KenjinxNative",
        )
        val path = rom.file.absolutePath
        for (cn in classes) {
            val cls = runCatching { Class.forName(cn) }.getOrNull() ?: continue
            val methods = try {
                cls.methods
            } catch (_: Throwable) {
                continue
            }
            for (m in methods) {
                val n = m.name
                if (!n.contains("Icon", true) && !n.contains("icon", true)) continue
                if (n.contains("GameInfo", true) || n.contains("GetGameInfo", true)) continue
                val raw = invokeIcon(m, path) ?: continue
                val bmp = when (raw) {
                    is ByteArray -> if (raw.size in 2000..400_000) BitmapFactory.decodeByteArray(raw, 0, raw.size) else null
                    is Bitmap -> raw
                    else -> null
                }
                if (bmp != null && bmp.width >= 64) {
                    BootLog.add("cover native ${cls.simpleName}.${m.name}")
                    return bmp
                }
            }
        }
        return null
    }

    private fun invokeIcon(m: java.lang.reflect.Method, path: String): Any? {
        return try {
            m.isAccessible = true
            val static = java.lang.reflect.Modifier.isStatic(m.modifiers)
            val obj: Any? = if (static) null else return null
            when (m.parameterTypes.size) {
                0 -> m.invoke(obj)
                1 -> {
                    val p = m.parameterTypes[0]
                    if (p == String::class.java) m.invoke(obj, path) else null
                }
                2 -> {
                    val a = m.parameterTypes[0]
                    val b = m.parameterTypes[1]
                    if (a == String::class.java && b == String::class.java) {
                        val ext = path.substringAfterLast('.', "nsp")
                        m.invoke(obj, path, ext)
                    } else null
                }
                else -> null
            }
        } catch (_: Throwable) {
            null
        }
    }

    private fun pfs0Icon(f: File): Bitmap? {
        RandomAccessFile(f, "r").use { raf ->
            if (raf.length() < 64) return null
            val magic = ByteArray(4)
            raf.readFully(magic)
            if (String(magic) != "PFS0") return null
            val hdr = ByteArray(12)
            raf.readFully(hdr)
            val le = ByteBuffer.wrap(hdr).order(ByteOrder.LITTLE_ENDIAN)
            val count = le.int
            val strSz = le.int
            if (count !in 1..80 || strSz !in 1..16384) return null
            val table = ByteArray(count * 0x18)
            raf.readFully(table)
            val names = ByteArray(strSz)
            raf.readFully(names)
            val data0 = raf.filePointer
            data class Ent(val off: Long, val size: Long)
            val ents = ArrayList<Ent>()
            for (i in 0 until count) {
                val e = ByteBuffer.wrap(table, i * 0x18, 0x18).order(ByteOrder.LITTLE_ENDIAN)
                val off = e.long
                val size = e.long
                if (size in 20_000L..2_500_000L) ents.add(Ent(data0 + off, size))
            }
            ents.sortBy { it.size }
            for (ent in ents.take(10)) {
                jpegScan(raf, ent.off, ent.size)?.let { return it }
            }
        }
        return null
    }

    private fun nroIcon(f: File): Bitmap? {
        RandomAccessFile(f, "r").use { raf ->
            return jpegScan(raf, 0, minOf(raf.length(), 3_000_000L))
        }
    }

    private fun jpegScan(raf: RandomAccessFile, start: Long, len: Long): Bitmap? {
        val n = len.coerceAtMost(2_500_000L).toInt()
        if (n < 2000) return null
        val buf = ByteArray(n)
        raf.seek(start)
        val got = raf.read(buf)
        if (got < 2000) return null
        var i = 0
        while (i < got - 4) {
            if (buf[i] == 0xFF.toByte() && buf[i + 1] == 0xD8.toByte() && buf[i + 2] == 0xFF.toByte()) {
                var j = i + 3
                while (j < got - 1) {
                    if (buf[j] == 0xFF.toByte() && buf[j + 1] == 0xD9.toByte()) {
                        val size = j + 2 - i
                        if (size in 2000..250_000) {
                            val bmp = BitmapFactory.decodeByteArray(buf, i, size)
                            if (bmp != null && bmp.width >= 64 && bmp.height >= 64) return bmp
                        }
                        break
                    }
                    j++
                }
            }
            i++
        }
        return null
    }

    private fun decodeFile(f: File): Bitmap? {
        return try {
            val opts = BitmapFactory.Options()
            opts.inJustDecodeBounds = true
            BitmapFactory.decodeFile(f.absolutePath, opts)
            if (opts.outWidth < 32 || opts.outHeight < 32) return null
            opts.inJustDecodeBounds = false
            opts.inSampleSize = if (opts.outWidth > 512) 2 else 1
            BitmapFactory.decodeFile(f.absolutePath, opts)
        } catch (_: Throwable) {
            null
        }
    }

    private fun save(f: File, bmp: Bitmap) {
        try {
            f.parentFile?.mkdirs()
            f.outputStream().use { bmp.compress(Bitmap.CompressFormat.JPEG, 85, it) }
        } catch (_: Throwable) {
        }
    }
}
